### PR TITLE
Expression language support for Authentication 

### PR DESCRIPTION
**Pull request by wetted**

### CODE REVIEW

File: security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java [hunk 1]
```diff
@@ -65,6 +68,16 @@ public Publisher<SecurityRuleResult> check(HttpRequest<?> request, @Nullable Rou
...
+                for (AnnotationValue<Secured> securedAnnotation: methodRoute.getAnnotationValuesByType(Secured.class)) {
+                    if (securedAnnotation instanceof EvaluatedAnnotationValue<Secured>) {
+                        Optional<SecurityRuleResult> result = securedAnnotation.booleanValue()
+                            .map(b -> (Boolean.TRUE.equals(b) ? SecurityRuleResult.ALLOWED : SecurityRuleResult.REJECTED));
+                        if (result.isPresent()) {
+                            return Mono.just(result.get());
+                        }
+                    }
+                }
                 Optional<String[]> optionalValue = methodRoute.getValue(Secured.class, String[].class);
```
Comment (graemerocher) [inline security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java:81]: not part of this PR but this should be changed to `methodRoute.stringValues(Secured.class)` as this call is less efficient

File: security-annotations/src/main/java/io/micronaut/security/annotation/expressions/SecuredEvaluationContextRegistrar.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Registers Security Evaluation Context.
+ * @author Dean Wette
+ * @since 4.0.0
+ */
+public class SecuredEvaluationContextRegistrar implements ExpressionEvaluationContextRegistrar {
+    @Override
+    public String getContextClassName() {
+        return "io.micronaut.security.expressions.SecuredEvaluationContext";
```
[the remark this answers is not in the document]
Comment (wetted) [reply security-annotations/src/main/java/io/micronaut/security/annotation/expressions/SecuredEvaluationContextRegistrar.java:28]: That's what I had to begin with but had teething issues getting it to work. Now that we have everything wired up I'll try adding that back.

File: security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java (hunk 1 above)
Comment (wetted) [reply security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java]: @graemerocher I made the change here since it made more sense as part of the 'non-repeatable' correction you mentioned above.

**Comment by wetted** [comment]
There's a bug that results in test failures for the Java and Kotlin docs examples. I'm investigating. I think I know what the cause is. I'm also still working on the docs for this.

File: security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java [hunk 3]
```diff
@@ -71,11 +71,20 @@ public Publisher<SecurityRuleResult> check(HttpRequest<?> request, @Nullable Rou
...
+//                        return Mono.just(result.get());
+//                    }
+
+                    // I think we want something like this. It's not exactly right, but I'll fix,
+                    // ...or we should consider adding to Secured: `String condition() default "";`
+                    // then we can keep the above code using `securedAnnotation.booleanValue("cpndition")`
+                    SecurityRuleResult result = Arrays.stream(securedAnnotation.stringValues())
+                        .anyMatch(Boolean::valueOf) ? SecurityRuleResult.ALLOWED : SecurityRuleResult.REJECTED;
+                        return Mono.just(result);
                 }
```
Comment (graemerocher) [inline security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java:88]: yeah this looks ok

File: security-annotations/src/main/java/io/micronaut/security/annotation/expressions/SecuredEvaluationContextRegistrar.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sdelamo) [reply security-annotations/src/main/java/io/micronaut/security/annotation/expressions/SecuredEvaluationContextRegistrar.java:28]: It was intentional to register the `SecuredEvaluationContext` globally.  I could see someone who wants to use the security context in a custom annotation, for example for ACL.

File: security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java (hunk 3 above)
Comment (sdelamo) [reply security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java:79]: I changed it to use `booleanValues()`

File: test-suite-kotlin/src/test/kotlin/io/micronaut/security/docs/securedexpressions/ExampleController.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import java.security.Principal
+
+@Requires(property = "spec.name", value = "docexpressions")
+//tag::exampleControllerExpressions[]
+@Controller("/authenticated")
+class ExampleController {
+    @Secured("#{ user?.attributes?.get('email') == 'sherlock@micronaut.example' }")
+    @Produces(MediaType.TEXT_PLAIN)
+    @Get("/email")
+    fun authenticationByEmail(principal: Principal) = "${principal.name} is authenticated"
```
Comment (sdelamo) [inline test-suite-kotlin/src/test/kotlin/io/micronaut/security/docs/securedexpressions/ExampleController.kt:18]: @wetted I have changed this to be more kotlin idiomatic

**Comment by sdelamo** [comment]
@wetted I have removed `getPrincipal` from Security expressions context. I thought about it and I don't think it make sense if we are already exposing `user` which is a super set of principal. 

@wetted I added docs as well. 

@graemerocher I think this is ready for review.

File: security/src/main/java/io/micronaut/security/expressions/SecuredEvaluationContext.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.security.authentication.Authentication;
+import io.micronaut.security.utils.SecurityService;
+import jakarta.inject.Singleton;
+
+/**
+ * Context for supporting {@link io.micronaut.security.annotation.Secured} annotation expressions with object references.
+ */
+@Singleton
+public class SecuredEvaluationContext {
```
Comment (graemerocher) [inline security/src/main/java/io/micronaut/security/expressions/SecuredEvaluationContext.java:27]: annotate with `@Internal` and make `final`

File: security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java [hunk 6]
```diff
@@ -65,6 +68,16 @@ public Publisher<SecurityRuleResult> check(HttpRequest<?> request, @Nullable Rou
         if (routeMatch instanceof MethodBasedRouteMatch) {
             MethodBasedRouteMatch<?, ?> methodRoute = ((MethodBasedRouteMatch) routeMatch);
             if (methodRoute.hasAnnotation(Secured.class)) {
+
+                for (AnnotationValue<Secured> securedAnnotation: methodRoute.getAnnotationValuesByType(Secured.class)) {
```
[the remark this answers is not in the document]
Comment (graemerocher) [reply security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java:71]: This is marked as resolved but it is not resolved

File: security/src/main/java/io/micronaut/security/expressions/SecuredEvaluationContext.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    public SecuredEvaluationContext(SecurityService securityService) {
+        this.securityService = securityService;
+    }
+
+    /**
+     * @return the authenticated user {@link io.micronaut.security.authentication.Authentication} or {@literal null}.
+     */
+    @Nullable
+    public Authentication getUser() {
```
Comment (graemerocher) [inline security/src/main/java/io/micronaut/security/expressions/SecuredEvaluationContext.java:39]: why `user` and not `authentication`?

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java (hunk 1 above)
Comment (graemerocher) [reply security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java:79]: I don't see these changes

Comment (sdelamo) [reply security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java]: resolved via https://github.com/micronaut-projects/micronaut-security/pull/1309/commits/d9287702f262ed23433eb14f3021ed5f402512be

File: security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sdelamo) [reply security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java]: resolved via https://github.com/micronaut-projects/micronaut-security/pull/1309/commits/00eb9d9ff10d842f99ec38ce3ee99dd5c411e544

File: security/src/main/java/io/micronaut/security/expressions/SecuredEvaluationContext.java (hunk 5 above)
Comment (sdelamo) [reply security/src/main/java/io/micronaut/security/expressions/SecuredEvaluationContext.java:29]: resolved via https://github.com/micronaut-projects/micronaut-security/pull/1309/commits/6935e79452ba1aa1f8f6dc3aad517781f0e1c231

**Comment by sdelamo** [comment]
@graemerocher I had to revert https://github.com/micronaut-projects/micronaut-security/pull/1309/commits/423b9178ae66e97d03f80c2c98d8be29c4bb1d21


```
@Controller("/example")
@Secured(SecurityRule.IS_AUTHENTICATED) 
public class ExampleController {

    @Produces(MediaType.TEXT_PLAIN)
    @Get("/admin")
    @Secured({"ROLE_ADMIN", "ROLE_X"}) 
```

```
String[] valuesArr = methodRoute.stringValues(Secured.class);
```

returns `["ROLE_ADMIN", "ROLE_X", "isAuthenticated"]`

**Review by graemerocher — APPROVED** [review]
